File: model/Wallet.js
import CardType from './CardType.js';
import Card from './Card.js';
import Money from './Money.js';
import Pocket from './Pocket.js';

class Wallet {
    constructor() {
        this._pockets = Array.from({ length: 5 }, (_, i) => new Pocket(i + 1));
    }

    putObjectIn(pocketNumber, obj) {
        const matchingPockets = this._pockets.filter(pocket => pocket.getNumber() === pocketNumber);
        if (matchingPockets.length > 0) {
            matchingPockets[0].putObject(obj);
        }
        return matchingPockets;
    }

    putObjectOut(objectId) {
        const obj = this.getObjects().find(o => o.getId() === objectId);
        if (obj) {
            this._pockets.forEach(pocket => pocket.retrieveObject(obj));
        }
        return this._pockets;
    }

    getObjectLocation(objectId) {
        const obj = this.getObjects().find(o => o.getId() === objectId);
        return obj ? this._pockets.find(pocket => pocket.getObjects().includes(obj)) : null;
    }

    getObjects() {
        return this._pockets.flatMap(pocket => pocket.getObjects());
    }

    getObjectIn(pocketNumber) {
        const matchingPocket = this._pockets.find(pocket => pocket.getNumber() === pocketNumber);
        return matchingPocket ? matchingPocket.getObjects() : [];
    }

    getBalance() {
        const money = this.getObjects().filter(obj => obj instanceof Money);
        return money.reduce((total, m) => total + m.getAmount(), 0);
    }

    countObject(objectType) {
        const cards = this.getObjects().filter(obj => obj instanceof Card);
        switch (objectType) {
            case CardType.NI_CARD:
                return cards.filter(cd => cd.getType() === CardType.NI_CARD).length;
            case CardType.CREDIT_CARD:
                return cards.filter(cd => cd.getType() === CardType.CREDIT_CARD).length;
            case CardType.DRIVING_CARD:
                return cards.filter(cd => cd.getType() === CardType.DRIVING_CARD).length;
            case CardType.VISIT_CARD:
                return cards.filter(cd => cd.getType() === CardType.VISIT_CARD).length;
            default:
                return 0;
        }
    }
}

export default Wallet;

File: model/Wallet.cs
using System;
using System.Collections.Generic;
using System.Linq;


public class Wallet
{
    private List<Pocket> pockets;

    public Wallet()
    {
        pockets = new List<Pocket>
        {
            new Pocket(1),
            new Pocket(2),
            new Pocket(3),
            new Pocket(4),
            new Pocket(5)
        };
    }

    public List<Pocket> PutObjectIn(int pocketNumber, PocketObject obj)
    {
        return pockets
            .Where(pk => pk.number == pocketNumber)
            .Peek(pocket => pocket.PutObject(obj))
            .ToList();
    }

    public List<Pocket> PutObjectOut(int objectId)
    {
        PocketObject obj = GetObjects().First(o => o.id == objectId);
        return pockets
            .Peek(pk => pk.RetrieveObject(obj))
            .ToList();
    }

    public Pocket GetObjectLocation(int objectId)
    {
        PocketObject obj = GetObjects().First(o => o.Id == objectId);
        return pockets.First(pk => pk.objects.Contains(obj));
    }

    public List<PocketObject> GetObjects()
    {
        List<PocketObject> objects = new List<PocketObject>();
        pockets.SelectMany(pk => pk.objects).ToList().ForEach(objects.Add);
        return objects;
    }

    public List<PocketObject> GetObjectIn(int pocketNumber)
    {
        return pockets.First(pk => pk.number == pocketNumber).objects;
    }

    public double GetBalance()
    {
        List<Money> money = new List<Money>();
        pockets.SelectMany(pk => pk.objects)
            .OfType<Money>()
            .ToList()
            .ForEach(obj => money.Add(obj));

        return money.Select(m => m.amount).Sum();
    }

    public int CountObject(CardType objectType)
    {
        List<Card> cards = GetObjects().OfType<Card>().ToList();

        return objectType switch
        {
            CardType.NI_CARD => cards.Count(cd => cd.Type == CardType.NI_CARD),
            CardType.CREDIT_CARD => cards.Count(cd => cd.Type == CardType.CREDIT_CARD),
            CardType.DRIVING_CARD => cards.Count(cd => cd.Type == CardType.DRIVING_CARD),
            CardType.VISIT_CARD => cards.Count(cd => cd.Type == CardType.VISIT_CARD),
            _ => 0
        };
    }
}